modules/development/utils/queue.utils.php

```
@@ -0,0 +1,110 @@
+<?php
+
+namespace UniEngine\Engine\Modules\Development\Utils;
+
+use UniEngine\Engine\Common\Exceptions;
+use UniEngine\Engine\Includes\Helpers\World\Resources;
+use UniEngine\Engine\Includes\Helpers\World\Elements;
+
+abstract class QueueType {
+    const Planetary = 'QueueType::Planetary';
+    const Technology = 'QueueType::Technology';
+}
+
+//  Arguments
+//      - $props (Object)
+//          - queue (Object)
+//              - type (QueueType)
+//              - content (Array<QueueElement>)
+//          - user (Object)
+//          - planet (Object)
+//
+//  Returns: Object
+//      - queuedResourcesToUse (Object<resourceKey: string, value: number>)
+//          How much resources are yet to be consumed by all of the queued
+//          elements. Does not count the first element of the queue, because these
+//          resources have already been "consumed".
+//      - queuedElementLevelModifiers (Object<elementID: string, levelModifier: number>)
+//          Difference between current state and the future state of queued element levels.
+//      - queuedElementsCount (Number)
+//          How many elements are there in queue.
+//
+function getQueueStateDetails ($props) {
+    $planet = &$props['planet'];
+    $user = &$props['user'];
+    $queue = $props['queue']['content'];
+    $queueType = $props['queue']['type'];
+
+    $resourceKeys = Resources\getKnownSpendableResourceKeys();
+    $queuedResourcesToUse = array_map(
+        function () { return 0; },
+        array_flip($resourceKeys)
+    );
+
+    $queuedElementLevelModifiers = [];
+
+    $objectToModifyLevels = [];
+
+    if ($queueType === QueueType::Planetary) {
+        $objectToModifyLevels = &$planet;
+    } else if ($queueType === QueueType::Technology) {
+        $objectToModifyLevels = &$user;
+    } else {
+        throw new Exceptions\UniEngineException("Invalid queue type ('{$queueType}')");
```

Should be moved up, to "fail fast" on props extraction.